### PR TITLE
[Merged by Bors] - chore(CategoryTheory/Sites): remove unused lemmas and simplify some proofs in Sites.Discrete

### PR DESCRIPTION
**Pull request by dagurtomas**
This PR does the following
- Renames `Sheaf.IsDiscrete` to `Sheaf.IsConstant` and changes the definition to being in the essential image of the constant sheaf functor instead of requiring the counit to be an isomorphism. 
- Removes two lemmas from the file `Sites.Discrete` which were essentially duplicates of other ones, and simplifies a proof to be able to remove an auxiliary lemma. 
- Moves the remaining contets of `Sites.Discrete` to the file `Sites.ConstantSheaf`.

---

I added these lemmas initially because I used them in #14027, but it turns out they're essentially duplicates of other lemmas, now merged into one lemma `constantSheafAdj_counit_w` which is used in this file and in #14027.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/CategoryTheory/Sites/IsConstant.lean [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+namespace CategoryTheory.Sheaf
+
+variable {C : Type*} [Category C] (J : GrothendieckTopology C) {A : Type*} [Category A]
+  [HasWeakSheafify J A] {t : C} (ht : IsTerminal t)
+
+/--
+A sheaf is constant if the counit of the constant sheaf adjunction applied to it is an isomorphism.
+-/
+abbrev IsConstant (F : Sheaf J A) : Prop := IsIso ((constantSheafAdj J A ht).counit.app F)
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sites/IsConstant.lean:49]: It would be better to ask that `F` belongs to the essential image of the functor `constantSheaf` (note also that it would not depend on the existence/a choice of a terminal object.). (If `C` corresponds to a topological space that is not connected, the definition `IsIso (....counit.app...)` is not the right definition of `IsConstant`.) Changing the definition should not be a problem for your applications to condensed objects.

File: Mathlib/CategoryTheory/Sites/IsConstant.lean [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+variable (A) in
+/--
+The constant sheaf functor commutes up to isomorphism with any equivalence of sheaf categories.
+
+This is an auxiliary definition used to prove `Sheaf.isConstant_iff_of_equivalence` below, which
+says that the property of a sheaf of being constant is invariant under equivalence of
+sheaf categories.
+-/
+noncomputable def equivCommuteConstant : let e : Sheaf J A ≌ Sheaf K A := sheafEquiv G J K A
+    constantSheaf J A ⋙ e.functor ≅ constantSheaf K A :=
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sites/IsConstant.lean:109]: This should probably be moved to `Sites.ConstantSheaf` or `Sites.Equivalence`. (I would suspect that there is a more general commutation of `constantSheaf` with `Functor.sheafPullback`, under suitable assumptions, but this could be left for a separate PR.)

Comment (dagurtomas) [reply Mathlib/CategoryTheory/Sites/IsConstant.lean]: Added a TODO

File: Mathlib/CategoryTheory/Sites/ConstantSheaf.lean [hunk 3]
```diff
@@ -45,22 +66,176 @@ sheafification of the constant presheaf.
...
-lemma constantSheafAdj_counit_app {T : C} (hT : IsTerminal T) (F : Sheaf J D) :
-    (constantSheafAdj J D hT).counit.app F =
-      (presheafToSheaf J D).map ((constantPresheafAdj D hT).counit.app F.val) ≫
-        (sheafificationAdjunction J D).counit.app F := by
+variable {D}
+
+namespace Sheaf
+
+/--
+A sheaf is constant if the counit of the constant sheaf adjunction applied to it is an isomorphism.
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sites/ConstantSheaf.lean:79]:
```suggestion
A sheaf is constant if it is in the essential image of the constant sheaf functor.
```

File: Mathlib/CategoryTheory/Sites/ConstantSheaf.lean [hunk 4]
```diff
@@ -45,22 +66,176 @@ sheafification of the constant presheaf.
...
+
+open IsDenseSubsite
+
+variable (D) in
+/--
+The constant sheaf functor commutes up to isomorphism the equivalence of sheaf categories induced
+by a dense subsite.
+-/
+noncomputable def equivCommuteConstant : let e : Sheaf J D ≌ Sheaf K D := sheafEquiv G J K D
+    constantSheaf J D ⋙ e.functor ≅ constantSheaf K D :=
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sites/ConstantSheaf.lean:143]: In the statement, it would be better to use `letI`, or replace e by `sheafEquiv ...`.

Comment (dagurtomas) [reply Mathlib/CategoryTheory/Sites/ConstantSheaf.lean:143]: I only used `let e := ...` because the name of `sheafEquiv` used to be so long :-) 
I'll change it to `sheafEquiv` in the statement.

File: Mathlib/CategoryTheory/Sites/ConstantSheaf.lean [hunk 5]
```diff
@@ -45,22 +66,171 @@ sheafification of the constant presheaf.
...
+
+instance (F : Sheaf J D) [HasTerminal C]
+    [IsIso <| (constantSheafAdj J D terminalIsTerminal).counit.app F] : IsConstant J F where
+  mem_essImage := ⟨_, ⟨asIso <| (constantSheafAdj J D terminalIsTerminal).counit.app F⟩⟩
+
+instance [(constantSheaf J D).Faithful] [(constantSheaf J D).Full] (F : Sheaf J D)
+    [IsConstant J F] {T : C} (hT : IsTerminal T) :
+    IsIso ((constantSheafAdj J D hT).counit.app F) := by
+  rw [isIso_counit_app_iff_mem_essImage]
+  exact F.mem_essImage_of_isConstant
```
Comment (joelriou) [inline Mathlib/CategoryTheory/Sites/ConstantSheaf.lean:108]: This looks like a loop. I would suggest keeping as an instance only the version `... [IsConstant J F] ... : IsIso ...`.
